proposals/ytt/004-schema-validation/README.md

```diff
@@ -0,0 +1,1569 @@
+#@   end
+#@ end
+
+#@data/values-schema
+#@schema/validation ("There should be exactly one (1) registry replica if Helm Charts are stored on the filesystem.", one_registry_if_pvc_is_filesystem)
+---
+persistence:
+  imageChartStorage:
+    type: filesystem
+  persistentVolumeClaim:
+    registry:
+      accessMode: ReadWriteOnce
+registry:
+  replicas: 2
+```
+
+Note: the `@ytt:data` module's `data.values` struct is _not_ populated until _after_ the Data Values Pre-Processing phase is complete.  As such, the `data.values` variable is not useful within validations.
+
+Authors are encouraged to minimize their use of this technique: it fixes the location of data values making it more expensive to reorganize them.
+
+#### Use Case: Validations over Nullables
+
+Whenever a node is annotated `@schema/nullable`, then its default value is `null`.
+
+```yaml
+#@data/values-schema
+---
+#@schema/validation min=42, max=42
+#@schema/nullable
+foo: 13
+```
+
+The author may opt to short-circuit validations when the actual value is `null`:
+
+```yaml
+#@data/values-schema
+---
+#@schema/validation min=42, max=42, when_null_skip=True
+#@schema/nullable
+foo: 13
+```
+
+(This is preferred instead of requiring _all_ validators to perform a null check to keep the overall complexity of the system to a minimum.)
+
+_(also considered: [automatically including `when_null_skip` when Data Value is nullable](#automatically-including-when_null_skip-for-nullable-data-values))._
+
+When the author has set the Data Value as "nullable" with the intent to require the consumer to supply their own value, this is the [Required Input (for types with no empty value)](#required-input-for-types-with-no-empty-value) use case.
+
+This keyword is a special case of [`when=`](#use-case-conditional-validations) because it is so common.
+
+#### Use Case: Conditional Validations
+
+The author can short-circuit validations under certain conditions:
+
+```yaml
+#@data/values-schema
+---
+#@schema/validation ("", valid_service_config), when=lambda v: v.enabled
+service:
+  enabled: true
+  type: NodePort
+```
+
+Before any validations are checked, the value of `when=` (itself being a validation) is checked.\
+If `false`, then none of the other validations are checked and the value is assumed "valid."
+
+Here, the contents of the `service:` are validated (via `valid_service_config()`) only if `service.enabled` is `true`.
+
+#### Use Case: Disable Validations
+
+A consumer can skip validating Data Values:
+
+```console
+$ ytt ... --data-value-schema-disable-validation
+```
+
+or the programmatic equivalent:
+
+```python
+lib.eval(data_value_schema_disable_validation=True)
+```
+
+Likewise, a consumer can skip validating output documents:
+
+```console
+$ ytt ... --output-disable-validation=True
+```
+
+or the programmatic equivalent:
+
+```python
+lib.eval(output_disable_validation=True)
+```
+
```

Comment:
@vrabbi asks:

> Any thought on allowing to skip a specific violation?
> Not sure what the syntax would be but i can see cases where when consuming third party packages i may want to disable a specific validation that isnt valid in my case.
> I don't think its a very common case but it would allow for a break glass solution that is still more secure then disabling all validations.

Reply:
At first blush, this starts to make more compelling the need for the ability for overlays to modify metadata in addition to structure (i.e. edit annotations, not just the YAML structure).... 🤔 

Curious if others have thoughts about this:
- can you name specific use cases where this would occur (or speculate as to how frequent this would be)?
- ideas on a syntax for doing this via the command-line?